### ISSUE TITLE
Debug-enabled builds?

### ISSUE THREAD

**Issue by warpdesign** [dosbox-staging/dosbox-staging#758]
- Are there dosbox-staging builds with debug enabled?
- If so, where can I find them?
- how do I bring up the debugger (classic appears to do that with ` Alt-Pause`: how do I do that if my computer doesn't have a pause key?)

**Comment by dreamer** [comment]
> * Are there dosbox-staging builds with debug enabled?
> * If so, where can I find them?

No, we do not provide pre-built versions of DOSBox with debugger feature enabled (neither the "debug" nor "heavy debug" flavours).

> * how do I bring up the debugger (classic DOSBox appears to do that with ` Alt-Pause` keys: how do I do that if my computer doesn't have a pause key?)

After building a version with debugger feature enabled for yourself, pressing Ctrl+F1 will bring up a keymapper, and in there a new action appears (called "Debugger") you will be able to remap this action to some other key combination, it will look like this:

![Screenshot from 2020-12-07 14-12-00](https://user-images.githubusercontent.com/3967/101355347-b9e56300-3896-11eb-925e-f11fdc00b78c.png)

*That being said, I consider the debugger feature to be broken currently and have no real incentive to fixing it*.

Contributions in this area are very welcome - starting with properly documenting how to build a version with Debugger feature enabled; quick documentation is as follows:

**Linux, macOS**: install ncurses development package using your native package manager, create a **release** build with additional configure flag, either: `--enable-debug` or `--enable-debug=heavy`. 

**Windows**: install pdcurses using vcpkg, manually edit file `src/platform/visualc/config.h` and enable the version of debugger you want, build **Release** configuration.

*Don't ask me what's the difference between debugger and heavy-debugger because authors didn't bother to write documentation. Remember about creating release build, otherwise our normal debug logs will interfere with debugger feature - confusing? Yes, yes it is :/*

**Keeping this issue open until we'll have proper documentation for building with this feature enabled.**

**Comment by warpdesign** [comment]
Thanks for the answer: I'll try to make a build for macOS.

Here is the information I found about the heavier debug mode (note that it was published more than **15 years ago**: not sure it's still accurate):

- Memory breakpoints that activates when the value of a memory location is changed.
- CPU log that outputs the next n commands executed by the CPU to a textfile.
- Heavylog: Dumps the last (currently 20000) executed commands to a textfile when a program crashes or interrupt CD is called
- Unlike the debug version, heavydebug does not use INT 3 to realize breakpoints. This way, they cannot be removed. So if you want to use breakpoints, heavydebug is the better choice.

see: https://www.vogons.org/viewtopic.php?t=3944

> that being said, I consider the debugger feature to be broken currently and have no real incentive to fixing it.
Any idea what's broken?

**Comment by dreamer** [comment]
From my very brief use:

- Debugger text UI randomly breaks and moves parts of the content up or down
- Normal logs break UI sometimes (it definitely happens in debug builds, but I think I've seen it in release builds as well)
- Switch between emulator and debugger ncurses UI triggers additional window events, and this might lead to debugger locking down
- Debugger regularly leaves the terminal in inconsistent state, forcing user to close the terminal window
- ncurses gui reacts badly to resizing

So combining text mode debugger UI resembling DOS debuggers from 90s and all the issues above - personally I find this feature unusable at the moment. Fixes and improvements will be very welcome :)
